src/dyncfg/src/lib.rs

```diff
@@ -155,16 +155,37 @@ impl<T: ConfigType> Config<T> {
 
 /// A type usable as a [Config].
 pub trait ConfigType: Into<ConfigVal> + Clone + Sized {
-    /// A const-compatible type suitable for use as the default value of configs
-    /// of this type.
-    type Default: Into<Self> + Clone;
-
     /// Converts a type-erased enum value to this type.
     ///
     /// Panics if the enum's variant does not match this type.
     fn from_val(val: ConfigVal) -> Self;
 }
 
+/// A trait for a type that can be converted into a type that implements
+/// [`ConfigType`].
+pub trait IntoConfigType: Clone {
```

Comment:
I'd be inclined to call this something more like `ConfigDefault`

Reply:
Ah, the reason I went with this name is because `ConfigSet::add` also uses this trait outside the context of setting the default value for a config:

https://github.com/MaterializeInc/materialize/pull/26470/files#diff-b4b292283b42b29bc3cb36b83ee22eebd4726fcc8ca0ea425dc2fbdf0fa92f28R381-R385

Does that change your opinion at all? I don't feel strongly at all!

Reply:
> > For my understanding: That `ConfigUpdates:add` generality isn't necessary to get json working, right? It's a QOL improvement? AFAICT all the `ConfigUpdates:add` calls happen in a non-const context where we can cough up a json object.
> 
> Yes, not necessary for JSON! `ConfigUpdates::add` currently takes `val: T`, and the way to preserve that was to adjust the bound to use `IntoConfigType`. We could, if we wanted, delete that method entirely and force all existing callers to use `add_dynamic` instead.

Sorry if I'm being dense, but wouldn't `val: T::ConfigType` work?

Reply:
> > Assuming the above is correct, I'd call the ConfigUpdates:add usage more incidental and something like ConfigDefault is slightly more descriptive of the primary motivation for adding the trait.
> 
> But this works for me!

Done!

Reply:
> Sorry if I'm being dense, but wouldn't val: T::ConfigType work?

It's overly constraining! The `T` is the type of the default value now, which for strings is often `&'static str`. And then `ConfigUpdates::add` would lock you in to using string literals for that config value, even though a `String` is totally usable in that context.

Reply:
(Ftr, this convo doesn't need to block the PR, I'm just confused)

I'm not sure I follow! If `ConfigUpdates::add` were `pub fn add<T: ConfigDefault>(&mut self, config: &Config<T>, val: T::ConfigType)`, then the val would be a `String` not a `&'static str` no?

```
    impl ConfigDefault for &str {
        type ConfigType = String;
```

Reply:
Oh, yes, sorry, I think that would work!